shared/ssz/hash_cache_test.go

```diff
@@ -0,0 +1,182 @@
+			Uint:         uInt,
+			Int64Slice:   is,
+		}
+
+	}
+	return object
+}
+
+func TestHashKeyFn_OK(t *testing.T) {
+	mRoot := &root{
+		Hash: common.HexToHash("0x0123456"),
+	}
+
+	key, err := hashKeyFn(mRoot)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if key != mRoot.Hash.Hex() {
+		t.Errorf("Incorrect hash key: %s, expected %s", key, mRoot.Hash.Hex())
+	}
+}
+
+func TestHashKeyFn_InvalidObj(t *testing.T) {
+	_, err := hashKeyFn("bad")
+	if err != ErrNotMerkleRoot {
+		t.Errorf("Expected error %v, got %v", ErrNotMerkleRoot, err)
+	}
+}
+
+func TestObjCache_byHash(t *testing.T) {
+	cache := newHashCache()
+
+	byteSl := [][]byte{{0, 0}, {1, 1}}
+	mr, err := merkleHash(byteSl)
+	if err != nil {
+		t.Fatal(err)
+	}
+	hs, err := hashedEncoding(reflect.ValueOf(byteSl))
+	if err != nil {
+		t.Fatal(err)
+	}
+	exists, _, err := cache.RootByEncodedHash(bytesutil.ToBytes32(hs))
+
+	if err != nil {
+		t.Fatal(err)
+	}
+	if exists {
+		t.Error("Expected block info not to exist in empty cache")
+	}
+
+	if _, err := cache.MerkleHashCached(byteSl); err != nil {
+		t.Fatal(err)
+	}
+
+	exists, fetchedInfo, err := cache.RootByEncodedHash(bytesutil.ToBytes32(hs))
+
+	if err != nil {
+		t.Fatal(err)
+	}
+	if !exists {
+		t.Error("Expected blockInfo to exist")
+	}
+	if bytes.Compare(mr, fetchedInfo.MarkleRoot) != 0 {
+		t.Errorf(
+			"Expected fetched info number to be %v, got %v",
+			mr,
+			fetchedInfo.MarkleRoot,
+		)
+	}
+	if fetchedInfo.Hash != bytesutil.ToBytes32(hs) {
+		t.Errorf(
+			"Expected fetched info hash to be %v, got %v",
+			hs,
+			fetchedInfo.Hash,
+		)
+	}
+}
+
+func TestMerkleHashWithCache(t *testing.T) {
+	cache := newHashCache()
+	for i := 0; i < 200; i++ {
+
+		runMerkleHashTests(t, func(val [][]byte) ([]byte, error) {
+			return merkleHash(val)
+		})
+
+	}
+
+	for i := 0; i < 200; i++ {
+
+		runMerkleHashTests(t, func(val [][]byte) ([]byte, error) {
+			return cache.MerkleHashCached(val)
+		})
+
+	}
+
+}
+
+func BenchmarkHashWithoutCache(b *testing.B) {
+	featureconfig.FeatureConfig().CacheTreeHash = false
+	First := GenerateJunkObject(100)
+	TreeHash(&tree{First: First, Second: First})
+	for n := 0; n < b.N; n++ {
+		TreeHash(&tree{First: First, Second: First})
+	}
+}
+
+func BenchmarkHashWithCache(b *testing.B) {
+	featureconfig.FeatureConfig().CacheTreeHash = true
+	First := GenerateJunkObject(100)
+	type tree struct {
+		First  []*JunkObject
+		Second []*JunkObject
+	}
+	TreeHash(&tree{First: First, Second: First})
+	for n := 0; n < b.N; n++ {
+		TreeHash(&tree{First: First, Second: First})
+	}
+}
+
+func TestBlockCache_maxSize(t *testing.T) {
+	maxCacheSize = 10000
+	cache := newHashCache()
+	for i := uint64(0); i < uint64(maxCacheSize+1025); i++ {
+
+		if err := cache.AddRoot(bytesutil.ToBytes32(bytesutil.Bytes4(i)), []byte{1}); err != nil {
+			t.Fatal(err)
+		}
+	}
+	log.Printf(
+		"hash cache key size is %d, itemcount is %d",
+		maxCacheSize,
+		cache.hashCache.ItemCount(),
+	)
+	time.Sleep(time.Second)
```

Comment:
```suggestion
	time.Sleep(1* time.Second)
```

I know it is the same behavior, but it reads better.